Added handling of axes to regular plotting functions

philippjfr opened the pull request:
This PR makes it possible to pass an existing matplotlib ``Axes`` object to the various plotting functions in the plotting module. I realize that in PR #1112 drawing directly onto the axes was disabled, instead always drawing to the axis that is currently open using the ``plt`` function calls. While that is a good fix for users using the ``plt`` interface, which is useful for simple usage, when building more complex plots the object-oriented approach is the way to go. This is crucial for some of the functionality we're building as part of the [cube-explorer](https://github.com/CubeBrowser/cube-explorer) project integrating Iris functionality within [HoloViews](http://holoviews.org).

It's a fairly small set of changes but not being familiar with the code I may have missed something. I think only providing this functionality in plot but not quickplot makes sense since the latter is suited towards usage in an interactive interpreter anyway. I'd be happy to add examples, documentation and tests once we agree these are the correct changes.

marqh commented on lib/iris/plot.py:
```diff
@@ -278,7 +278,8 @@ def _draw_2d_from_bounds(draw_method_name, cube, *args, **kwargs):
         u, v = plot_arrays
         u, v = _broadcast_2d(u, v)
-        draw_method = getattr(plt, draw_method_name)
+        axes = kwargs.pop('axes')
```
there will need to be a None returned if axes doesn't exist

philippjfr replied:
Thanks for catching that, will make any fixes once you're done reviewing.

marqh commented:
you also need to change the copyright statement in the first line of plot.py

it needs to read 
```
# (C) British Crown Copyright 2010 - 2016, Met Office
```
as you've changed it in 2016

philippjfr commented:
Thanks @marqh, will do that and will hopefully get the contributor agreement sorted out quickly. Any thoughts/recommendations on examples, documentation and tests?

marqh commented:
i recommend 
https://github.com/SciTools/iris/tree/master/lib/iris/tests/unit/plot
as the start point for testing

the unit tests (lib/iris/tests/unit/.../*.py) are gradually replacing the older tests (lib/iris/tests/*.py)

marqh commented:
I think a lot of the plot types in
http://scitools.org.uk/iris/docs/latest/iris/iris/plot.html

will want this behaviour

I think that we should explicitly mention 'axes' on any plot type that we have implemented the change for
e.g.
https://github.com/philippjfr/iris/blob/master/lib/iris/plot.py#L615

philippjfr commented:
>I think that we should explicitly mention 'axes' on any plot type that we have implemented the change for.

Agreed, I think they should all support it now, so I'll just make sure to document the kwarg everywhere.

marqh commented:
CLA received, with thanks
https://github.com/SciTools/scitools.org.uk/pull/120

marqh commented:
Hi @philippjfr 

this looks grand.  please add a stub to
https://github.com/philippjfr/iris/tree/master/docs/iris/src/whatsnew/contributions_1.10
and squish your commits

then this should be good to merge
thank you

philippjfr commented:
@marqh Thanks for the review, I'll make sure to do that. I've also looked into adding some unit tests based on the existing unit tests but as far as I can tell only the contour and contourf calls are actually setting the x- and y-ticks correctly, so give me some time to investigate that.

philippjfr commented:
@marqh Made some final fixes, added unit tests, added an entry under contributions and squashed the commits. I think it's ready to merge, but let me know if you want me to refactor those unit tests a little bit.

pelson commented on lib/iris/plot.py:
```diff
@@ -529,7 +533,8 @@ def _ensure_cartopy_axes_and_determine_kwargs(x_coord, y_coord, kwargs):
     # Ensure the current axes are a cartopy.mpl.GeoAxes instance.
-    _replace_axes_with_cartopy_axes(cartopy_proj)
+    if not kwargs.get('axes'):
+        _replace_axes_with_cartopy_axes(cartopy_proj)
```
I'm not sure we can make this conditional - what if you've made an axes which isn't a cartopy GeoAxes and passed that through... it may be that we need to raise an exception in that situation.

philippjfr replied:
Good point, raising an exception makes sense to me.

philippjfr replied:
Raising an exception now.

pelson commented on lib/iris/plot.py:
```diff
@@ -610,14 +617,19 @@ def contour(cube, *args, **kwargs):
+        axis of the plot and the second element is the vertical axis
+        of the plot.
+
+    * axes: :class:`cartopy.mpl.geoaxes.GeoAxes` instance to call the
```
This isn't necessarily true: it is very often a normal matplotlib axes.

philippjfr replied:
As you say, as long as I raise an exception when it's expecting a GeoAxes I think we can safely change all these to regular matplotlib axes.

pelson commented on lib/iris/plot.py:
```diff
@@ -630,14 +642,19 @@ def contourf(cube, *args, **kwargs):
 
-    See :func:`matplotlib.pyplot.contourf` for details of other valid keyword
-    arguments.
+    * axes: :class:`cartopy.mpl.geoaxes.GeoAxes` instance to call the
```
Ditto

pelson commented on lib/iris/plot.py:
```diff
@@ -870,14 +910,19 @@ def points(cube, *args, **kwargs):
+        axis of the plot and the second element is the vertical axis
+        of the plot.
+
+    * axes: :class:`cartopy.mpl.geoaxes.GeoAxes` instance to call the
```
Maybe we can change all of these docstrings to:

```
 * axes: the :class:`matplotlib.axes.Axes` to use for drawing.
```

philippjfr replied:
Done.

pelson commented on lib/iris/plot.py:
```diff
@@ -1036,11 +1092,16 @@ def citation(text, figure=None):
         anchor.patch.set_boxstyle('round, pad=0, rounding_size=0.2')
-        figure.gca().add_artist(anchor)
+        axes = axes if axes else figure.gca()
+        axes.add_artist(anchor)
```
I don't think this needs to be axes - it could just as well be ```figure.patches.append(anchor)``` (untested, but along those lines). If that is the case, then axes is simply not needed as a kwarg.

philippjfr replied:
You can probably do that but would you want to control whether the text gets attached to an axis or the overall figure?

pelson replied:
That is reasonable for this PR. :+1:

philippjfr commented:
@pelson I've finally made the revisions you asked for, except regarding your comment to attach the anchored text to the figure. I'd be happy to revert adding an axes argument there for for now, just let me know what you think is best.

pelson commented on lib/iris/plot.py:
```diff
@@ -529,7 +533,11 @@ def _ensure_cartopy_axes_and_determine_kwargs(x_coord, y_coord, kwargs):
-    _replace_axes_with_cartopy_axes(cartopy_proj)
+    if 'axes' not in kwargs:
+        _replace_axes_with_cartopy_axes(cartopy_proj)
+    elif not not isinstance(kwargs['axes'], cartopy.mpl.geoaxes.GeoAxes):
```
double not. I'm not not sure what happens here...

This highlights the need for a test of this error I think.

philippjfr replied:
Agreed, will add one now.

pelson replied:
Looks like some of the tests are failing because of this too:

```
======================================================================

ERROR: test_outline (iris.tests.test_mapping.TestLimitedAreaCube)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/tests/test_mapping.py", line 240, in test_outline

    iplt.outline(self.cube)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 845, in outline

    antialiased=True, coords=coords, axes=axes)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 244, in _draw_2d_from_bounds

    cube, plot_defn, *args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 614, in _map_common

    return plotfn(*new_args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/matplotlib/pyplot.py", line 2946, in pcolormesh

    ret = ax.pcolormesh(*args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/matplotlib/axes.py", line 7766, in pcolormesh

    t = t._as_mpl_transform(self.axes)

  File "lib/cartopy/_crs.pyx", line 213, in cartopy._crs.CRS._as_mpl_transform (lib/cartopy/_crs.c:4071)

ValueError: Axes should be an instance of GeoAxes, got <class 'matplotlib.axes.AxesSubplot'>

======================================================================

ERROR: test_grid (iris.tests.test_mapping.TestBoundedCube)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/tests/test_mapping.py", line 205, in test_grid

    iplt.outline(self.cube)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 845, in outline

    antialiased=True, coords=coords, axes=axes)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 244, in _draw_2d_from_bounds

    cube, plot_defn, *args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/Iris-1.10.0.dev0-py2.7-linux-x86_64.egg/iris/plot.py", line 614, in _map_common

    return plotfn(*new_args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/matplotlib/pyplot.py", line 2946, in pcolormesh

    ret = ax.pcolormesh(*args, **kwargs)

  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/matplotlib/axes.py", line 7766, in pcolormesh

    t = t._as_mpl_transform(self.axes)

  File "lib/cartopy/_crs.pyx", line 213, in cartopy._crs.CRS._as_mpl_transform (lib/cartopy/_crs.c:4071)

ValueError: Axes should be an instance of GeoAxes, got <class 'matplotlib.axes.AxesSubplot'>
```

philippjfr replied:
Likely due to the double not above. Fixing that and adding new unit tests now.

pelson commented on lib/iris/plot.py:
```diff
@@ -601,7 +609,9 @@ def _map_common(draw_method_name, arg_func, mode, cube, plot_defn,
     # Draw the contour lines/filled contours.
-    return getattr(plt, draw_method_name)(*new_args, **kwargs)
+    axes = kwargs.pop('axes', None)
+    plotfn = getattr(axes if axes else plt, draw_method_name)
```
technically, this could be ```axes or plt``` - I'm happy enough with the ternary form though.

philippjfr commented:
Fixed the double ``not`` bug above, which will hopefully fix the tests and added some further unit tests to ensure that plots actually raise the GeoAxes exception when passed a regular axes.

philippjfr commented:
Seems there are further exceptions being raised. I don't quite understand how that's happening unless those tests have been passing axes in all along and they were simply ignored until now. Will investigate.

philippjfr commented:
Resolved the failing tests, the logic for raising an exception was off. I think it's ready to merge now.

pelson commented:
Thanks @philippjfr - this is an excellent addition. :+1:

philippjfr commented:
Yay! Thanks for merging @pelson.
